### PR TITLE
Binary Situations

### PR DESCRIPTION
**Pull request by Shadowfiend**
'Bout to slice and dice this sucka.

```
Supported flags:
    --debug
        Enable debug output.

    --rpc <rpc-url>
        Set RPC URL to the specified value.

    --mnemonic <mnemonic>
        Use the specified for the operating account. Also supports private key
        strings, since the underlying provider accepts these.

    --account <account>
        Use the specified account for all transactions. If --mnemonic is
        specified, it must be able to sign for this account in order for
        mutating transactions to be sent to the Ethereum chain. If this is
        left off, the first account for the private key is used.

Supported commands:
    deposit new [--no-mint] <lot-size-satoshis>
        Initiates a deposit funding flow. Takes the lot size in satoshis.
        Will prompt with a Bitcoin address when funding needs to be
        submitted. When the flow completes, outputs the deposit as a single
        tab-delimited line with the deposit address, current deposit state,
        the deposit lot size in satoshis, and, when applicable, the minted
        amount of TBTC.

        --no-mint
            Specifies not to mint TBTC once the deposit is qualified.

    deposit list [--vending-machine] [--address <address>]
        With no options, lists the deposits currently owned by the web3
        account address. Deposits are output as tab-delimited lines that
        include the deposit address, current deposit state, and deposit
        lot size in satoshis.

        --vending-machine
            Lists the deposits currently owned by the vending machine.

        --address <address>
            Lists the deposits currently owned by the specified address.

    deposit <address> [<resume|redeem|liquidate|withdraw>]
        Operations on a particular address. If no command is provided,
        outputs the deposit as a single tab-delimited line with the deposit
        address, current deposit state, and deposit lot size in satoshis.

        resume [--funding|--redemption] [--no-mint]
            Resumes a funding or redemption flow, depending on the deposit's
            current state. When the flow completes, outputs the deposit as a
            single tab-delimited line with the deposit address, current
            deposit state, and deposit lot size in satoshis.

            --funding
                Only resumes the funding flow and outputs the final deposit
                state; if the deposit is not mid-funding, does not resume and
                outputs an error.

                --no-mint
                    When resuming a funding flow, if the deposit is not already
                    mid-minting, specifies not to mint TBTC once the deposit is
                    qualified.

            --redemption
                Only resumes a flow if it is a redemption flow and outputs
                the final deposit state; if the deposit is not mid-redemption,
                does not resume and outputs an error.

        redeem <bitcoin-address>
            Initiates a deposit redemption flow that will redeem the deposit's
            BTC to the specified Bitcoin address. When the flow completes,
            outputs the deposit as a single tab-delimited line with the
            deposit address, current deposit state, deposit lot size in
            satoshis, and the transaction hash of the redemption Bitcoin
            transaction.

        courtesy-call
            Attempts to notify the deposit it is undercollateralized and
            should transition into courtesy call.

        liquidate [--for <funding-timeout|undercollateralization|courtesy-timeout|redemption-timeout>]
            Attempts to liquidate the deposit, reporting back the status of
            the liquidation . By default, looks for any available reason to
            liquidate. When the flow completes, outputs the deposit as a
            single tab-delimited line with the deposit address, current
            deposit state, deposit lot size in satoshis, and the liquidation
            status (\`liquidated\`, \`in-auction\`, or \`failed\`).

            --for <funding-timeout|undercollateralization|courtesy-timeout|redemption-timeout>
                If specified, only triggers liquidation for the specified
                reason. If the reason does not apply, reports \`not-applicable\`
                status.

        withdraw [--dry-run]
            Attempts to withdraw the current account's allowance from a tBTC
            deposit. Only the amount allowed for the current account is
            withdrawn. Outputs the withdrawn amount in wei once withdrawal
            is complete.

            --dry-run
                Outputs the amount that would be withdrawn in wei, but does
                not broadcast the transaction to withdraw it.

    lot-sizes
      Returns a list of the currently available lot sizes, one per line.

    supply
        Returns the current supply as a decimal amount in TBTC. 18 decimals of
        precision, but with a decimal point.

    supply-cap
        Returns the current supply cap as a decimal amount in TBTC.
```

### CODE REVIEW

**Comment by Shadowfiend** [comment]
Going to break this up into proper commits tomorrow.

File: src/Deposit.js
Comment (afmsavage) [inline src/Deposit.js:1392]: Extra 'o' in Out?  Didn't pull code, only looking on Github.com so can't search, sorry if this is correct!

File: bin/commands/deposit.js
Comment (Shadowfiend) [reply bin/commands/deposit.js:121]: It's a `prettier` default to format this way, and what `prettier` says goes :)

File: src/Deposit.js
Comment (Shadowfiend) [reply src/Deposit.js:1392]: Whew. Yep.

Comment (Shadowfiend) [reply src/Deposit.js]: Fixed!

**Comment by Shadowfiend** [comment]
Reworked the PR situation, I think this will be ready for review once I get a green light from the build. Notably, I'm still testing the various functionalities, but they can be iterated upon in the current structure.

**Comment by mhluongo** [comment]
That conflict might've been me merging the other PR first and I'm so sorry :sweat:

**Comment by Shadowfiend** [comment]
Nope, `package-lock` was already conflicting, was waiting to resolve until I got an approval :)

**Review by mhluongo — APPROVED** [review]
I couldn't get `deposit.js` reviewed as well as I'd like but I think this is good enough, considering most of it isn't dApp-impacting at all

File: bin/helpers.js
Comment (Shadowfiend) [reply bin/helpers.js:82]: We'll probably add more lines to our package-lock than total lines of code in this helper lib heh.

File: src/Deposit.js
Comment (Shadowfiend) [reply src/Deposit.js:67]: Nope, apparently we don't 🙆‍♂️ Even TypeScript ones are unexpectedly problematic, as I understand it, though I haven't looked deep.

File: bin/tbtc.js
Comment (Shadowfiend) [reply bin/tbtc.js:84]: I'm basically checking we don't have any other flags that weren't consumed. Def something that can be abstracted out.

**Pull request merged**
